### PR TITLE
Address-expose locals under complex local addresses in block morphing

### PR DESCRIPTION
**Pull request by SingleAccretion**
In block morphing, `addrSpill` is used when the destination or source represent indirections of "complex" addresses.  unfortunately, some trees in the form of `IND(ADDR(LCL))` fall into this category.

If such an `ADDR(LCL)` is used as an `addrSpill`, the underlying local __must__ be marked as address-exposed. Block morphing was using a very simplistic test for when that needs to happen, essentially only recognizing `ADDR(LCL_VAR/FLD)`. But it is possible to have a more complicated pattern as `PrepareDst/Src` use `IsLocalAddrExpr` to recognize indirect stores to locals.

Currently it appears impossible to get a mismatch here as morph transforms `IND(ADD(ADDR(LCL_VAR), OFFSET))` into `LCL_FLD` (including for `TYP_STRUCT` indirections), but this is a very fragile invariant. Transforming `TYP_STRUCT` `GT_FIELD`s into  `GT_OBJ`s instead of `GT_IND`s breaks it, for example.

Fix this by address-exposing the local obtained via `IsLocalAddrExpr`.

[No diffs](https://dev.azure.com/dnceng/public/_build/results?buildId=1524953&view=ms.vss-build-web.run-extensions-tab) as expected.

### CODE REVIEW

File: src/coreclr/jit/morphblock.cpp [hunk 1]
```diff
@@ -1148,28 +1149,13 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
...
-    // "BaseAddress" method, that handles fields of structs, before or after
-    // morphing.
-    if ((addrSpill != nullptr) && addrSpill->OperIs(GT_ADDR))
-    {
-        GenTree* addrSpillOp = addrSpill->AsOp()->gtGetOp1();
-        if (addrSpillOp->IsLocal())
-        {
-            // We will *not* consider this to define the local, but rather have each individual field assign
-            // be a definition.
-            addrSpillOp->gtFlags &= ~(GTF_LIVENESS_MASK);
```
Comment (SingleAccretion) [inline src/coreclr/jit/morphblock.cpp:1167]: This was deleted as it was redundant with the code [above](https://github.com/dotnet/runtime/blob/8f80eed12d4f9846b3300ae97ff077d6c71746d7/src/coreclr/jit/morphblock.cpp#L1123-L1132).

File: src/coreclr/jit/morphblock.cpp [hunk 2]
```diff
@@ -1197,23 +1184,16 @@ GenTree* MorphCopyBlockHelper::CopyFieldByField()
...
+        if (addrSpillSrcLclNum != BAD_VAR_NUM)
         {
-            GenTree* addrOp = addrSpill->AsOp()->gtOp1;
-            if (addrOp->IsLocal())
-            {
-                unsigned lclVarNum = addrOp->AsLclVarCommon()->GetLclNum();
-                m_comp->lvaGetDesc(lclVarNum)->SetAddressExposed(true DEBUGARG(AddressExposedReason::COPY_FLD_BY_FLD));
-                m_comp->lvaSetVarDoNotEnregister(lclVarNum DEBUGARG(DoNotEnregisterReason::AddrExposed));
-            }
+            m_comp->lvaSetVarAddrExposed(addrSpillSrcLclNum DEBUGARG(AddressExposedReason::COPY_FLD_BY_FLD));
```
Comment (SingleAccretion) [inline src/coreclr/jit/morphblock.cpp:1189]: `lvaSetVarAddrExposed` sets `lvaSetVarDoNotEnregister`.

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

**Comment by AndyAyersMS** [comment]
> the underlying local must be marked as address-exposed.

We really should get out of the business of using address exposed to cover up weaknesses in later analyses. Did you look at trying to fix liveness?

**Comment by SingleAccretion** [comment]
> Did you look at trying to fix liveness?

Hmm, I don't think liveness is at fault here. It is by design that it (and SSA) do not recognize stores to spilled local addresses, the previous phases are supposed to turn them into `IND(ADDR(LCL))/LCL_FLD`s.

So this is a block morphing issue - why not always use `GT_LCL_FLD`? The reason is that we only use it when we have matching handles for source and destination - we attach field sequences using only source's field handles. We should instead use the field handles of the destination's type (if it is different) - there is actually an existing correctness issue associated with this, but it only shows up when `addrSpill` is a pointer that VN understands. Fixing that is part of #58312 and will be addressed.

However, it is separate from this fix, part of the work towards enabling `TYP_STRUCT` `LCL_FLD`, that just makes sure we use `addrSpill` "correctly" (even as it should not be used for local destinations/sources).

**Comment by AndyAyersMS** [comment]
So perhaps update this comment to reflect some of the above?

https://github.com/dotnet/runtime/blob/f04a24249835096eea1a1a66e4af03cfec5ed32b/src/coreclr/jit/morphblock.cpp#L1189-L1194

If we are stuck with this longer term, seems like we could annotate the spill temp somehow since we know where it points.

**Comment by AndyAyersMS** [comment]
Retrying some of the failed legs.

**Comment by AndyAyersMS** [comment]
There was a bad arm64 test machine -- retrying yet again.

**Review by AndyAyersMS — APPROVED** [review]
Changes LGTM.

Thank you.

**Pull request merged**